authority/ssh.go

```diff
@@ -146,22 +146,24 @@ func (a *Authority) GetSSHBastion(ctx context.Context, user, hostname string) (*
 }
 
 // SignSSH creates a signed SSH certificate with the given public key and options.
-func (a *Authority) SignSSH(_ context.Context, key ssh.PublicKey, opts provisioner.SignSSHOptions, signOpts ...provisioner.SignOption) (*ssh.Certificate, error) {
+func (a *Authority) SignSSH(_ context.Context, key ssh.PublicKey, opts provisioner.SignSSHOptions, signOpts ...provisioner.SignOption) (cert *ssh.Certificate, err error) {
```

Comment:
After looking at this package's dependents, I am willing to accept a breaking change that accepts a provisioner as a parameter or `Sign` and `SignSSH`. We can add the provisioner as a return value in `authority.Authorize`, and just pass it.

I'm not worried about breaking changes, but an option that won't cause less breaking changes would be to store the provisioner in the context, although we will need to add a context to `Sign` or just create the `SignContext` method. 

In any case, I'm ok to clean these, and have always a context and a provisioner parameter.

Reply:
If we do something like this we can kill a few issues at once:
```go
func (a *Authority) SignSSH(_ context.Context, key ssh.PublicKey, opts provisioner.SignSSHOptions, signOpts ...provisioner.SignOption (*ssg.Certificate, err error) {
   certs, prov, err = a.signSSH(...)
   a.incrProvisionerCounter(prov, err, Meter.SSHSigned) 
   return certs, err
}
```

We can keep the current interface, although, as I said, I'm not opposed to changing it. But especially, it will be harder to cause bugs in the future because a PR adds something like `if err := foo(); err != nil { ... }`, well and we won't need the pointer to an interface.

Reply:
Cleaning up sounds nice and I hope that we will. There's quite a few things that could benefit from being worked on in many places.

Having said that, named returns and 'defer' statements aren't uncommon when dealing with instrumented code and I don't see how adding yet another level of indirection is helpful.

If we instrument `SignSSH` and not `signSSH` then a caller of `signSSH` would need to remember to instrument the call.

The pointer to a pointer (interface), could be solved by refactoring the defer statement to be:

```go
defer func() { incrProvisionerCounter(prov, err, a.Meter.SSHSigned) }()
```

instead of 

```go
defer a.incrProvisionerCounter(prov, err, Meter.SSHSigned)
```

Other than verbosity, there's not a lot of difference to these signatures.

Reply:
@maraino Adding a `SignContext` sounds good. In fact, I have that in https://github.com/smallstep/certificates/pull/1542. I should have a look at that again, sometime soon 😅